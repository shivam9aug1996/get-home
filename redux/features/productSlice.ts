import { createSlice } from "@reduxjs/toolkit";
import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { baseUrl } from "../constants";

const normalizeProducts = (products: any) => {
  const productIds: any = [];
  const productsById: any = {};

  products.forEach((product: any) => {
    productIds.push(product._id);
    productsById[product._id] = product;
  });

  return { productIds, productsById };
};

export const productApi = createApi({
  reducerPath: "productApi",
  baseQuery: fetchBaseQuery({
    baseUrl: `${baseUrl}`,
  }),
  endpoints: (builder) => ({
    fetchProducts: builder.query({
      query: () => ({
        url: "/properties",
        method: "GET",
      }),
      transformResponse: (response: any) => {
        const normalizedData = normalizeProducts(response?.properties);
        return normalizedData;
      },
    }),
    // fetchAllProducts: builder.query({
    //   query: (data) => ({
    //     url: "/allProducts",
    //     method: "GET",
    //     params: {
    //       search: data?.search,
    //     },
    //   }),
    //   transformResponse: (response) => {
    //     const normalizedData = normalizeProducts(response?.products);
    //     return normalizedData;
    //   },
    // }),
  }),
});

const productSlice = createSlice({
  name: "productSlice",
  initialState: {},
  reducers: {},
  extraReducers: (builder) => {},
});

export const { useFetchProductsQuery } = productApi;

export default productSlice.reducer;
