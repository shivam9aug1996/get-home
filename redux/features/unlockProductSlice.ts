import { createSlice } from "@reduxjs/toolkit";
import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { baseUrl } from "../constants";

export const unlockProductApi = createApi({
  reducerPath: "unlockProductApi",
  baseQuery: fetchBaseQuery({
    baseUrl: `${baseUrl}`,
  }),
  tagTypes: ["unlockProducts"],
  endpoints: (builder) => ({
    fetchUnlockProducts: builder.query({
      query: (data) => ({
        url: "/properties/unlock",
        method: "GET",
        params: {
          userId: data?.userId,
        },
      }),
      providesTags: ["unlockProducts"],
    }),
    unlockProduct: builder.mutation({
      query: (data) => ({
        url: "/properties/unlock",
        method: "POST",
        body: data,
      }),
      invalidatesTags: ["unlockProducts"],
    }),
    // fetchAllProducts: builder.query({
    //   query: (data) => ({
    //     url: "/allProducts",
    //     method: "GET",
    //     params: {
    //       search: data?.search,
    //     },
    //   }),
    //   transformResponse: (response) => {
    //     const normalizedData = normalizeProducts(response?.products);
    //     return normalizedData;
    //   },
    // }),
  }),
});

const unlockProductSlice = createSlice({
  name: "unlockProductSlice",
  initialState: {
    currentProduct: { data: {}, isUnlocked: false },
  },
  reducers: {
    setProduct: (state, action) => {
      state.currentProduct = action.payload;
    },
  },
  extraReducers: (builder) => {},
});

export const { useFetchUnlockProductsQuery, useUnlockProductMutation } =
  unlockProductApi;
export const { setProduct } = unlockProductSlice.actions;

export default unlockProductSlice.reducer;
