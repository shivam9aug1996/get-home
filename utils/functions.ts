import * as Location from "expo-location";

// Utility function to validate mobile number
const isValidMobileNumber = (mobileNumber: string): string => {
  const mobileNumberRegex = /^[6-9]\d{9}$/;
  if (!mobileNumber.trim()) return "Please fill in mobile number.";
  if (!mobileNumber.match(mobileNumberRegex))
    return "Please enter a valid Indian mobile number.";
  if (mobileNumber.length !== 10)
    return "Mobile number should be 10 digits long.";
  return "";
};

// Utility function to validate password
const isValidPassword = (password: string): string => {
  if (!password.trim()) return "Please fill in password.";
  if (password.length < 4)
    return "Password should be at least 4 characters long.";
  return "";
};

// Function to validate login form
export const validateLoginForm = (
  mobileNumber: string,
  password: string
): string => {
  let error = isValidMobileNumber(mobileNumber);
  if (error) return error;

  error = isValidPassword(password);
  if (error) return error;

  return "";
};

// Function to validate signup form
export const validateSignupForm = (
  name: string,
  mobileNumber: string,
  password: string
): string => {
  if (!name.trim()) return "Please fill in your name.";

  let error = validateLoginForm(mobileNumber, password);
  if (error) return error;

  return "";
};

export const getDistance = (lat1: any, lon1: any, lat2: any, lon2: any) => {
  const toRadians = (degree: any) => (degree * Math.PI) / 180;
  const R = 6371e3; // Earth's radius in meters
  const φ1 = toRadians(lat1);
  const φ2 = toRadians(lat2);
  const Δφ = toRadians(lat2 - lat1);
  const Δλ = toRadians(lon2 - lon1);

  const a =
    Math.sin(Δφ / 2) * Math.sin(Δφ / 2) +
    Math.cos(φ1) * Math.cos(φ2) * Math.sin(Δλ / 2) * Math.sin(Δλ / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

  return R * c; // Distance in meters
};

export const getCurrentLocation = () => {
  let cachedLocation: Location.LocationObject | null = null;
  return async (showToast: any) => {
    if (!cachedLocation) {
      const { status } = await Location.requestForegroundPermissionsAsync();
      if (status !== "granted") {
        showToast("Permission to access location was denied");
        return null;
      }
      cachedLocation = await Location.getCurrentPositionAsync({});
    }
    return cachedLocation;
  };
};

export const getCachedLocation = getCurrentLocation();
