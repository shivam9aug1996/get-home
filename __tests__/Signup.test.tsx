import React from "react";

import { useSignupMutation as useSignupMutationOriginal } from "@/redux/features/authSlice";
import { fireEvent, render, waitFor } from "@testing-library/react-native";
import renderer from "react-test-renderer";
import { validateSignupForm } from "@/utils/functions";
import Toast from "react-native-root-toast";
import SignupScreen from "@/app/signup";

jest.mock("@/redux/features/authSlice", () => ({
  useSignupMutation: jest.fn(),
}));
jest.mock("@/utils/functions", () => ({
  validateSignupForm: jest.fn(),
}));
jest.mock("react-native-root-toast", () => ({
  show: jest.fn(),
  hide: jest.fn(),
  positions: {
    TOP: "top",
    BOTTOM: "bottom",
    CENTER: "center",
  },
}));

const useSignupMutation = useSignupMutationOriginal as jest.MockedFunction<
  typeof useSignupMutationOriginal
>;
describe("SignupScreen", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  useSignupMutation.mockReturnValue([
    jest.fn(),
    {
      isLoading: false,
      reset: jest.fn(),
      error: undefined,
      isSuccess: false,
      isError: false,
      data: null,
      originalArgs: undefined,
    },
  ]);

  it("should render signup screen correctly", () => {
    const tree = renderer.create(<SignupScreen />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should show validation error toast if login credentials are invalid", async () => {
    const { getByTestId } = render(<SignupScreen />);
    (validateSignupForm as jest.Mock).mockReturnValue("Invalid credentials");

    fireEvent.press(getByTestId("signup-button"));
    await waitFor(() => {
      expect(Toast.show).toHaveBeenCalledWith("Invalid credentials", {
        position: Toast.positions.TOP,
      });
    });
  });
});
