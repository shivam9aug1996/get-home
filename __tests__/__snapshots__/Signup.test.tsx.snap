// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupScreen should render signup screen correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
      }
    }
  >
    <Text
      style={
        {
          "marginTop": 25,
        }
      }
    >
      Signup
    </Text>
    <View
      style={
        {
          "marginTop": 20,
          "width": "80%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="Name"
        placeholderTextColor="#11181C"
        style={
          [
            {
              "borderColor": "#ccc",
              "borderRadius": 5,
              "borderWidth": 1,
              "fontSize": 16,
              "marginBottom": 10,
              "paddingHorizontal": 12,
              "paddingVertical": 10,
            },
            {
              "color": "#11181C",
            },
            undefined,
          ]
        }
        value=""
      />
      <TextInput
        keyboardType="phone-pad"
        maxLength={10}
        onChangeText={[Function]}
        placeholder="Mobile Number"
        placeholderTextColor="#11181C"
        style={
          [
            {
              "borderColor": "#ccc",
              "borderRadius": 5,
              "borderWidth": 1,
              "fontSize": 16,
              "marginBottom": 10,
              "paddingHorizontal": 12,
              "paddingVertical": 10,
            },
            {
              "color": "#11181C",
            },
            undefined,
          ]
        }
        textContentType="telephoneNumber"
        value=""
      />
      <TextInput
        onChangeText={[Function]}
        placeholder="Password"
        placeholderTextColor="#11181C"
        secureTextEntry={true}
        style={
          [
            {
              "borderColor": "#ccc",
              "borderRadius": 5,
              "borderWidth": 1,
              "fontSize": 16,
              "marginBottom": 10,
              "paddingHorizontal": 12,
              "paddingVertical": 10,
            },
            {
              "color": "#11181C",
            },
            undefined,
          ]
        }
        textContentType="password"
        value=""
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#a679a6",
            "borderRadius": 10,
            "marginTop": 5,
            "opacity": 1,
            "paddingVertical": 10,
            "width": "100%",
          }
        }
        testID="signup-button"
      >
        <Text
          style={
            {
              "color": "#fff",
              "fontSize": 18,
              "textAlign": "center",
            }
          }
        >
          Signup
        </Text>
      </View>
    </View>
  </View>
</View>
`;
